src/crimson/common/errorator.h

```diff
@@ -0,0 +1,185 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <exception>
+
+namespace ceph {
+
+namespace _impl {
+  enum class ct_error {
+    enoent,
+    invarg,
+    enodata
+  };
+}
+
+// unthrowable_wrapper ensures compilation failure when somebody
+// would like to `throw make_error<...>)()` instead of returning.
+// returning allows for the compile-time verification of future's
+// AllowedErrorsV and also avoid the burden of throwing.
+template <_impl::ct_error ErrorV> struct unthrowable_wrapper {
+  unthrowable_wrapper(const unthrowable_wrapper&) = delete;
+  static constexpr unthrowable_wrapper instance{};
+  template <class T> friend const T& make_error();
+
+  // comparison operator for the static_assert in future's ctor.
+  template <_impl::ct_error OtherErrorV>
+  constexpr bool operator==(const unthrowable_wrapper<OtherErrorV>&) const {
+    return OtherErrorV == ErrorV;
+  }
+
+private:
+  // can be used only for initialize the `instance` member
+  explicit unthrowable_wrapper() = default;
+};
+
+template <class T> [[nodiscard]] const T& make_error() {
+  return T::instance;
+}
+
+template <class... WrappedAllowedErrorsT>
+struct errorator {
+  template <class... ValuesT>
```

Comment:
probably we can discourage using of variadic template? as seastar decided to deprecate this feature.

Reply:
As the `errorator<...>::future` derives from `seastar::future`, we get the warnings out-of-the-box. When the future is removed, we can turn `... ValuesT` into `ValueT`.